src/module_3/module3_statistical_learning.md

```diff
@@ -0,0 +1,1368 @@
+    if validation_size > 0:
+        train_order_ids, validation_order_ids = train_test_split(train_order_ids,
+                                                                 test_size=validation_size / (1 - test_size),
+                                                                 random_state=random_state)
+
+    # Create split masks
+    train_mask = df['order_id'].isin(train_order_ids)
+    validation_mask = df['order_id'].isin(validation_order_ids)
+    test_mask = df['order_id'].isin(test_order_ids)
+
+    # Split the DataFrame based on masks
+    train_df = df[train_mask]
+    validation_df = df[validation_mask]
+    test_df = df[test_mask]
+
+    return train_df, validation_df, test_df
+
+train_df_0, validation_df_0, test_df_0 = split_by_order_id(df_orders)
+```
+
+### Create a baseline model
+
+
+```python
+train_df = train_df_0.copy()
+validation_df = validation_df_0.copy()
+
+### Label Encoder
+encoders = {}
+for col in categorical_cols:
+    le = LabelEncoder()
+    train_df[col] = le.fit_transform(train_df[col])
+    encoders[col] = le
+
+# Save the encoding to reuse it in validation and test
+with open(data_path + 'label_encoders.pkl', 'wb') as f:
+    pickle.dump(encoders, f)
+
+# ------------------ Model Training ------------------
+    
+X_train = train_df.drop(label_col, axis=1)
+y_train = train_df[label_col]
+
+# Fit the LogisticRegression model
+log_reg = LogisticRegression() 
+log_reg.fit(X_train, y_train) 
+
+# ------------------ Classification ------------------
+
+# Encode test data
+with open(data_path + 'label_encoders.pkl', 'rb') as f:
+    encoders = pickle.load(f)
+
+for col in categorical_cols: 
+    validation_df[col] = encoders[col].transform(validation_df[col])
+
+X_val = validation_df.drop(label_col, axis=1)
+y_val = validation_df[label_col]
+
+# Make predictions
+y_pred = log_reg.predict(X_val) 
+
+```
+
+
+```python
+def model_performance_metrics(y_test, y_pred):
+    accuracy = accuracy_score(y_test, y_pred)
+    conf_matrix = confusion_matrix(y_test, y_pred, labels=[1, 0])
+    precision = precision_score(y_test, y_pred, labels=[1, 0])
+    recall = recall_score(y_test, y_pred, labels=[1, 0])
+    f1 = f1_score(y_test, y_pred)
+    roc_auc = roc_auc_score(y_test, y_pred)
+
+    print(f'Accuracy: {accuracy}')
+    print(f'Confusion Matrix:\n {conf_matrix}')
+    print(f'Precision: {precision}')
+    print(f'Recall: {recall}')
+    print(f'F1-Score: {f1}')
+    print(f'ROC AUC: {roc_auc}')
+
+model_performance_metrics(y_val, y_pred)
+```
+
+    Accuracy: 0.9854730694473871
+    Confusion Matrix:
+     [[     0   3154]
+     [     0 213960]]
+    Precision: 0.0
+    Recall: 0.0
+    F1-Score: 0.0
+    ROC AUC: 0.5
+
+
+Como se podía esperar, el modelo predice sólo 0s ya que las clases están totalmente desbalanceadas. El siguiente paso será aplicar un undersampling para igualar las clases
+
+## Undersampling
+
+En este paso voy a aplicar un undersampling para tener las clases balanceadas al 50%
```

Comment:
what's the point of this? As commented in the class undersampling si useless in most of the cases unless your classes are very unbalanced (which is not the case)